tests/xoto3/dynamodb/write_versioned/example_unit_test.py

```diff
@@ -0,0 +1,76 @@
+"""Imagine you had the following business logic and you wanted to test
+it. Here's how you might go about that, using the
+xoto3.dynamodb.write_versioned API itself to set everything up and
+prove that it worked.
+
+The example here is the same as the code we have in the README.
+"""
+import pytest
+
+import xoto3.dynamodb.write_versioned as wv
+
+# threadsafe because they are pure values.
+task_table = wv.ItemTable("TaskTable", item_name="Task")
+user_table = wv.ItemTable("UserTable", item_name="User")
+
+# imagine these were passed in via an API
+task_key = dict(id="task1")
+new_task = dict(task_key, name="plant tree", num_subtasks=3, user_id="steve")
+
+# this transaction builder is a pure function that constructs
+# a value representing DynamoDB effects to be transactionally/atomically applied.
+def create_and_link_task_unless_exists(t: wv.VersionedTransaction) -> wv.VersionedTransaction:
+    t = task_table.presume(task_key, None)(t)
+    # ^ we presume that the task does not exist, i.e. has the value None
+    # This is a no-op if a value has already been fetched.
+    # This is also purely an optimization to avoid an initial read;
+    # the whole transaction would result in the same data in the table without it.
+
+    existing_task = task_table.get(task_key)(t)
+    if existing_task:
+        return t
+        # ^ perform no action as long as the item already exists
+
+    t = task_table.put(new_task)(t)
+    # ^ put this item into this table as long as the item does not exist
+    user = user_table.require(dict(id=new_task["user_id"]))(t)  # type: ignore
+    # ^ fetch the user and raise an exception if it does not exist
+    user["task_ids"].append(task_key["id"])
+    t = user_table.put(user)(t)
+    # ^ make sure that the user knows about its task
+    return t
+    # ^ return the built transaction to be executed.
+
+
+def test_task_and_user_are_not_written_if_task_already_exists():
+    t = wv.VersionedTransaction(dict())
+    existing_task = dict(task_key, name="water flowers", is_done=True, user_id="felicity")
+    t = task_table.presume(task_key, existing_task)(t)
+    # presume lets us set up a fixture in the fake table, represented by the VersionedTransaction
+
+    t = create_and_link_task_unless_exists(t)
+
+    resulting_task = task_table.require(task_key)(t)
+    assert resulting_task == existing_task
+    # the task remains unchanged
+
+    with pytest.raises(wv.ItemNotFetchedException):
+        user_table.get(dict(id="felicity"))(t)
+        # the user was not even fetched, much less written
+
+
+def test_task_and_user_are_both_written_if_task_does_not_exist():
+    t = wv.VersionedTransaction(dict())
+    t = task_table.presume(task_key, None)(t)
+    # declare that the task does not exist - essentially this is a 'fixture', but without I/O
+    user_key = dict(id="steve")
+    user = dict(user_key, name="Actually Steve", task_ids=list())
+    t = user_table.presume(user_key, user)(t)
```

Comment:
Just like the above question, I believe this is because `versioned_transact_write_items` has not been called with the transaction that `create_and_link_task_unless_exists` resulted in.

Reply:
presume never fails. from the docstring: "Returns a modified transaction if the value of the item is not already known."

it used to be even more explicit that this was a no-op if the value is already known. i suppose I'll want to go back to that.

Reply:
I'm sorry. I didn't mean that the actual call to `presume` would fail. I just mean that calling it with data that doesn't actually reflect the real data in the database would end up causing a failure when committing the transaction. 

I thought that was what made this more than just some `initial_state` function.

Reply:
we're not talking to an actual database in a database. Since we're never actually _executing_ the transaction in our unit test, all we're testing is the pure business logic of "what will the state of the database be when the transaction is successfully executed?"

remember, while it's technically true that an incorrect presumption will make your transaction fail, that is transparent to the transaction builder, which will always just get called again and be presented with the correct state of the database rather than their incorrect presumption. In other words, it's not a transaction failure, it's a transparent transaction retry. Assuming your transaction builder is a pure function with no side effects, of course :)